README.md
# ts-transformers-plantain

Some custom typescript transformers.

[![Dependency Status](https://david-dm.org/plantain-00/ts-transformers-plantain.svg)](https://david-dm.org/plantain-00/ts-transformers-plantain)
[![devDependency Status](https://david-dm.org/plantain-00/ts-transformers-plantain/dev-status.svg)](https://david-dm.org/plantain-00/ts-transformers-plantain#info=devDependencies)
[![Build Status: Linux](https://travis-ci.org/plantain-00/ts-transformers-plantain.svg?branch=master)](https://travis-ci.org/plantain-00/ts-transformers-plantain)
[![Build Status: Windows](https://ci.appveyor.com/api/projects/status/github/plantain-00/ts-transformers-plantain?branch=master&svg=true)](https://ci.appveyor.com/project/plantain-00/ts-transformers-plantain/branch/master)
![Github CI](https://github.com/plantain-00/ts-transformers-plantain/workflows/Github%20CI/badge.svg)
[![npm version](https://badge.fury.io/js/ts-transformers-plantain.svg)](https://badge.fury.io/js/ts-transformers-plantain)
[![Downloads](https://img.shields.io/npm/dm/ts-transformers-plantain.svg)](https://www.npmjs.com/package/ts-transformers-plantain)
[![gzip size](https://img.badgesize.io/https://unpkg.com/ts-transformers-plantain?compression=gzip)](https://unpkg.com/ts-transformers-plantain)
[![type-coverage](https://img.shields.io/badge/dynamic/json.svg?label=type-coverage&prefix=%E2%89%A5&suffix=%&query=$.typeCoverage.atLeast&uri=https%3A%2F%2Fraw.githubusercontent.com%2Fplantain-00%2Fts-transformers-plantain%2Fmaster%2Fpackage.json)](https://github.com/plantain-00/ts-transformers-plantain)

## install

`yarn add ts-transformers-plantain`

## features

### executedCodeTransformer

For every statements, prepend a console log of file name and line number, so when the statement is executed, the file name and line number will be logged in console.

for example:

```txt
[executed code]demo/index.tsx:1:1
[executed code]demo/index.tsx:2:3
[executed code]demo/index.tsx:3:3
```

logger can be disabled by `executed-code:disable` in comment.
children logger can be disabled by `executed-code:disable-children` in comment.

### codeTimeTransformer

For example:

```txt
[code time]demo/index.tsx:3:3: 1ms
[code time]demo/index.tsx:3:3: 1ms
[code time]demo/index.tsx:1:1: 16ms
```

logger can be disabled by `code-time:disable` in comment.
children logger can be disabled by `code-time:disable-children` in comment.

`codeTimeTransformerFactory({ threshold: 1000 })` can be used to log only when code time >= the threshold (`>=1000ms` for this example)

### lastExecutingCodeTransformer

It can be used to record last executing code before exception, it's useful for environment without stacktrace or clear error message, eg, ExtendScript.

For example:

```txt
demo/index.tsx:16:3
```

logger can be disabled by `last-executing-code:disable` in comment.
children logger can be disabled by `last-executing-code:disable-children` in comment

### codeMemoryNodeJsTransformer

For example:

```txt
[code memory nodejs]demo/index.tsx:10:3: +0.02MB -> 2.02MB
[code memory nodejs]demo/index.tsx:10:3: +0.01MB -> 2.35MB
[code memory nodejs]demo/index.tsx:10:3: +0.01MB -> 2.62MB
[code memory nodejs]demo/index.tsx:10:3: +0.01MB -> 2.63MB
[code memory nodejs]demo/index.tsx:10:3: +0.01MB -> 2.64MB
[code memory nodejs]demo/index.tsx:10:3: +0.01MB -> 2.65MB
[code memory nodejs]demo/index.tsx:10:3: +0.01MB -> 2.67MB
[code memory nodejs]demo/index.tsx:10:3: +0.01MB -> 2.68MB
[code memory nodejs]demo/index.tsx:4:1: +0.71MB -> 2.71MB
```

logger can be disabled by `code-memory-nodejs:disable` in comment.
children logger can be disabled by `code-memory-nodejs:disable-children` in comment.

### codeMemoryBrowserTransformer

For example:

```txt
[code memory browser]demo/index.tsx:10:3: +0.02MB -> 2.02MB
[code memory browser]demo/index.tsx:10:3: +0.01MB -> 2.35MB
[code memory browser]demo/index.tsx:10:3: +0.01MB -> 2.62MB
[code memory browser]demo/index.tsx:10:3: +0.01MB -> 2.63MB
[code memory browser]demo/index.tsx:10:3: +0.01MB -> 2.64MB
[code memory browser]demo/index.tsx:10:3: +0.01MB -> 2.65MB
[code memory browser]demo/index.tsx:10:3: +0.01MB -> 2.67MB
[code memory browser]demo/index.tsx:10:3: +0.01MB -> 2.68MB
[code memory browser]demo/index.tsx:4:1: +0.71MB -> 2.71MB
```

logger can be disabled by `code-memory-browser:disable` in comment.
children logger can be disabled by `code-memory-browser:disable-children` in comment.

### executedCodeCoverageTransformer

This transformer will collect all executed code line number, `_showCodeCoverage()` will print the result in console, eg:

```txt
[executed code coverage]demo/index.tsx: 12 / 21 57.14% 27,31,32,34,39,42,45,48,55
[executed code coverage]total: 12 / 21 57.14%
```

### jsxElementPositionTransformer

This transformer will add code position in `data-_position` for all jsx elements, for example:

```txt
React.createElement("button", { onClick: () => console.info(1), "data-_position": 'demo/index.tsx:55:1' });
```

## safeOptimizationTransformer

This transformer will do some safe optiomizaiton, for example:

```ts
const a = 'aa'.length // -> const a = 2
```

## usage

### with ttypescript

```txt
"plugins": [
  {
    "transform": "ts-transformers-plantain",
    "import": "executedCodeTransformer",
    "type": "raw"
  },
]
```

### with ts-loader

```txt
import { executedCodeTransformer } from "ts-transformers-plantain"

{
  test: /\.ts$/,
  loader: 'ts-loader',
  options: {
    transpileOnly: true,
    getCustomTransformers: (program) => ({
      before: [executedCodeTransformer]
    })
  }
}
```


CHANGELOG.md
# Change Log

## [1.10.0](https://github.com/plantain-00/ts-transformers-plantain/compare/v1.9.0...v1.10.0) (2020-08-29)
  
* feat: support codeTimeTransformerFactory ([6246e88](https://github.com/plantain-00/ts-transformers-plantain/commit/6246e88fab1f90ac7bd0aef6cf47565cd13786a0))

## [1.9.0](https://github.com/plantain-00/ts-transformers-plantain/compare/v1.8.0...v1.9.0) (2020-08-29)
  
* feat: support typescript 4 ([0412dde](https://github.com/plantain-00/ts-transformers-plantain/commit/0412ddef89e6d7af5af8659fe96bb83a5be657e1))
* fix: add missing parserOptions ([59e3f75](https://github.com/plantain-00/ts-transformers-plantain/commit/59e3f75dedd30e1d20db4f18ce7dc13254c08730))
* chore: update dependencies ([f75dab1](https://github.com/plantain-00/ts-transformers-plantain/commit/f75dab1f0cbabb001f8526c30839baa9cf9277e2)) ([b590a23](https://github.com/plantain-00/ts-transformers-plantain/commit/b590a2342b7d75f28515be8031039fb9290fa342)) ([06dacb3](https://github.com/plantain-00/ts-transformers-plantain/commit/06dacb3d76c72dc67149f2eb15a4c61c82cc455f))
* build: improve release script ([ede0a3c](https://github.com/plantain-00/ts-transformers-plantain/commit/ede0a3c170b0a78c6135aa66fcad51ecb75eba72))

## [1.8.0](https://github.com/plantain-00/ts-transformers-plantain/compare/v1.7.0...v1.8.0) (2020-07-26)
  
* fix: release script ([f063afe](https://github.com/plantain-00/ts-transformers-plantain/commit/f063afee1181fd32bb45391d51c050814fc2522f))
* feat: support safeOptimizationTransformer ([a28d37f](https://github.com/plantain-00/ts-transformers-plantain/commit/a28d37f51e9d03cfc5b8be98286e2dd75723b2b5))
* chore: update dependencies ([54959d6](https://github.com/plantain-00/ts-transformers-plantain/commit/54959d6150ff8275a610c7ffc3f25fd7efc14ec8)) ([b1d9bee](https://github.com/plantain-00/ts-transformers-plantain/commit/b1d9bee38dd92313567a4dcdb2add2be5c1adf02)) ([38aada5](https://github.com/plantain-00/ts-transformers-plantain/commit/38aada53afa5a4b56c7d9da3b4185374103ab2ca)) ([3be68f0](https://github.com/plantain-00/ts-transformers-plantain/commit/3be68f09b313e80e4372a19b35bbccd8a17f3949)) ([115929a](https://github.com/plantain-00/ts-transformers-plantain/commit/115929a60bb2479afd3b3f003e61b97bc190e13d)) ([c80f522](https://github.com/plantain-00/ts-transformers-plantain/commit/c80f5222d6881137817899aff96a79ffaa1f1a79)) ([c9e8eeb](https://github.com/plantain-00/ts-transformers-plantain/commit/c9e8eeba2a4251d35399605f32bb8ff2ac91ee7b)) ([300d22e](https://github.com/plantain-00/ts-transformers-plantain/commit/300d22ecb564ac29eb183f9dd2843a9a23589612))
* chore: add changelog ([90ca2f8](https://github.com/plantain-00/ts-transformers-plantain/commit/90ca2f8909dcb9da296fbde64dcc4d9f085ff52a))
* docs: add github ci badge ([d79e37d](https://github.com/plantain-00/ts-transformers-plantain/commit/d79e37d9d33dac166c06a0ee7d0d5a7b9a6fdb3f))
* chore: remove tsconfig.eslint.json ([0ab1da1](https://github.com/plantain-00/ts-transformers-plantain/commit/0ab1da1e1764f65258e9e2ea1c483480c539a80c))
* chore: reuse tsconfig.json ([eb58619](https://github.com/plantain-00/ts-transformers-plantain/commit/eb586194d4616fac1e8b0be8d2847b92c91f06ae))
* chore: remove commitlint ([a2c3cb2](https://github.com/plantain-00/ts-transformers-plantain/commit/a2c3cb285e8d68f6bf18fa553bbd3c8ab3dae553))

## [1.7.0](https://github.com/plantain-00/ts-transformers-plantain/compare/v1.6.0...v1.7.0) (2020-05-27)
  
* feat: add jsx element position ([13f684e](https://github.com/plantain-00/ts-transformers-plantain/commit/13f684e627b99bbde72499be608f00a7a65c1f31))

## [1.6.0](https://github.com/plantain-00/ts-transformers-plantain/compare/v1.5.0...v1.6.0) (2020-05-26)
  
* refactor: improve memory log ([589c013](https://github.com/plantain-00/ts-transformers-plantain/commit/589c0136d04d3c6c7cd1938342206a748413883c))
* feat: code memory log increased and total memory ([78f4601](https://github.com/plantain-00/ts-transformers-plantain/commit/78f4601aa3b7971c1a30d08d75bc1f4946464b03))

## [1.5.0](https://github.com/plantain-00/ts-transformers-plantain/compare/v1.4.0...v1.5.0) (2020-05-24)
  
* feat: support executedCodeCoverageTransformer ([6ef8af9](https://github.com/plantain-00/ts-transformers-plantain/commit/6ef8af91898456d549eab4426501cead414ba5d3))

## [1.4.0](https://github.com/plantain-00/ts-transformers-plantain/compare/v1.3.0...v1.4.0) (2020-05-24)
  
* feat: support memory transformer ([a0d04a7](https://github.com/plantain-00/ts-transformers-plantain/commit/a0d04a730202222ca488a69016e499d7c412abde))
* feat: support disable children and show function or class method name ([0a376f6](https://github.com/plantain-00/ts-transformers-plantain/commit/0a376f6026f0ec3dbda41f011e49e93f52d4a113))

demo/index2.js
/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

"use strict";
/* WEBPACK VAR INJECTION */(function(global) {
const _global = typeof globalThis !== "undefined" ? globalThis : typeof window !== "undefined" ? window : global;
if (!_global._executedCode) {
    _global._executedCode = new Map();
    _global._allCode = new Map();
    _global._showCodeCoverage = function () {
        var total = 0;
        var totalExecuted = 0;
        _global._allCode.forEach(function (all, file) {
            var set = _global._executedCode.get(file);
            var remain = [];
            for (var j = 0; j < all.length; j++) {
                var e = all[j];
                total++;
                if (set.has(e)) {
                    totalExecuted++;
                }
                else {
                    remain.push(e);
                }
            }
            remain.sort(function (a, b) { return a - b; });
            var percent = Math.floor(10000 * set.size / all.length) / 100;
            var percentString = percent.toFixed(2);
            console.log("[executed code coverage]" + file + ": " + set.size + " / " + all.length + " " + percentString + "% " + remain.join(","));
        });
        var percent = Math.floor(10000 * totalExecuted / total) / 100;
        var percentString = percent.toFixed(2);
        console.log("[executed code coverage]total: " + totalExecuted + " / " + total + " " + percentString + "%");
    };
}
var _executedCodeLines = new Set();
_global._executedCode.set("demo/index.tsx", _executedCodeLines);
_global._allCode.set("demo/index.tsx", [2, 4, 15, 26, 30, 37, 53, 55, 57, 58, 5, 10, 11, 16, 19, 27, 31, 34, 32, 39, 42, 45, 48]);
console.debug("[executed code]demo/index.tsx:2:1");
var _code_time_0 = Date.now();
var _code_memory_browser_0 = performance.memory.usedJSHeapSize;
_executedCodeLines.add(2);
// last-executing-code:disable
let _lastExecutingCode;
var _code_memory_browser_after_0 = performance.memory.usedJSHeapSize;
var _code_memory_browser_changes_0 = _code_memory_browser_after_0 - _code_memory_browser_0;
if (_code_memory_browser_changes_0 > 10485.76) {
    console.debug("[code memory browser]demo/index.tsx:2:1: +" + Math.round(_code_memory_browser_changes_0 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_0 / 10485.76) / 100) + "MB");
}
if (Date.now() !== _code_time_0) {
    console.debug("[code time]demo/index.tsx:2:1: " + (Date.now() - _code_time_0) + "ms");
}
console.debug("[executed code]demo/index.tsx:4:1");
var _code_time_1 = Date.now();
_lastExecutingCode = "demo/index.tsx:4:1";
var _code_memory_browser_1 = performance.memory.usedJSHeapSize;
_executedCodeLines.add(4);
for (let i = 0; i < 4; i++) {
    console.debug("[executed code]demo/index.tsx:5:3");
    var _code_time_7 = Date.now();
    _lastExecutingCode = "demo/index.tsx:5:3";
    var _code_memory_browser_7 = performance.memory.usedJSHeapSize;
    _executedCodeLines.add(5);
    let j = i;
    var _code_memory_browser_after_7 = performance.memory.usedJSHeapSize;
    var _code_memory_browser_changes_7 = _code_memory_browser_after_7 - _code_memory_browser_7;
    if (_code_memory_browser_changes_7 > 10485.76) {
        console.debug("[code memory browser]demo/index.tsx:5:3: +" + Math.round(_code_memory_browser_changes_7 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_7 / 10485.76) / 100) + "MB");
    }
    if (Date.now() - _code_time_7 >= 1000) {
        console.debug("[code time]demo/index.tsx:5:3: " + (Date.now() - _code_time_7) + "ms");
    }
    _lastExecutingCode = "demo/index.tsx:10:3";
    var _code_memory_browser_8 = performance.memory.usedJSHeapSize;
    _executedCodeLines.add(10);
    /**
     * executed-code:disable
     * code-time:disable
     */
    while (j < 4) {
        _lastExecutingCode = "demo/index.tsx:11:5";
        var _code_memory_browser_9 = performance.memory.usedJSHeapSize;
        _executedCodeLines.add(11);
        j++;
        var _code_memory_browser_after_9 = performance.memory.usedJSHeapSize;
        var _code_memory_browser_changes_9 = _code_memory_browser_after_9 - _code_memory_browser_9;
        if (_code_memory_browser_changes_9 > 10485.76) {
            console.debug("[code memory browser]demo/index.tsx:11:5: +" + Math.round(_code_memory_browser_changes_9 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_9 / 10485.76) / 100) + "MB");
        }
    }
    var _code_memory_browser_after_8 = performance.memory.usedJSHeapSize;
    var _code_memory_browser_changes_8 = _code_memory_browser_after_8 - _code_memory_browser_8;
    if (_code_memory_browser_changes_8 > 10485.76) {
        console.debug("[code memory browser]demo/index.tsx:10:3: +" + Math.round(_code_memory_browser_changes_8 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_8 / 10485.76) / 100) + "MB");
    }
}
var _code_memory_browser_after_1 = performance.memory.usedJSHeapSize;
var _code_memory_browser_changes_1 = _code_memory_browser_after_1 - _code_memory_browser_1;
if (_code_memory_browser_changes_1 > 10485.76) {
    console.debug("[code memory browser]demo/index.tsx:4:1: +" + Math.round(_code_memory_browser_changes_1 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_1 / 10485.76) / 100) + "MB");
}
if (Date.now() !== _code_time_1) {
    console.debug("[code time]demo/index.tsx:4:1: " + (Date.now() - _code_time_1) + "ms");
}
console.debug("[executed code]demo/index.tsx:15:1");
var _code_time_2 = Date.now();
_lastExecutingCode = "demo/index.tsx:15:1";
var _code_memory_browser_2 = performance.memory.usedJSHeapSize;
_executedCodeLines.add(15);
try {
    console.debug("[executed code]demo/index.tsx:16:3");
    var _code_time_8 = Date.now();
    _lastExecutingCode = "demo/index.tsx:16:3";
    var _code_memory_browser_10 = performance.memory.usedJSHeapSize;
    _executedCodeLines.add(16);
    throw 1;
    var _code_memory_browser_after_10 = performance.memory.usedJSHeapSize;
    var _code_memory_browser_changes_10 = _code_memory_browser_after_10 - _code_memory_browser_10;
    if (_code_memory_browser_changes_10 > 10485.76) {
        console.debug("[code memory browser]demo/index.tsx:16:3: +" + Math.round(_code_memory_browser_changes_10 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_10 / 10485.76) / 100) + "MB");
    }
    if (Date.now() - _code_time_8 >= 1000) {
        console.debug("[code time]demo/index.tsx:16:3: " + (Date.now() - _code_time_8) + "ms");
    }
}
catch {
    console.debug("[executed code]demo/index.tsx:19:3");
    var _code_time_9 = Date.now();
    var _code_memory_browser_11 = performance.memory.usedJSHeapSize;
    _executedCodeLines.add(19);
    // last-executing-code:disable
    console.info(_lastExecutingCode);
    var _code_memory_browser_after_11 = performance.memory.usedJSHeapSize;
    var _code_memory_browser_changes_11 = _code_memory_browser_after_11 - _code_memory_browser_11;
    if (_code_memory_browser_changes_11 > 10485.76) {
        console.debug("[code memory browser]demo/index.tsx:19:3: +" + Math.round(_code_memory_browser_changes_11 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_11 / 10485.76) / 100) + "MB");
    }
    if (Date.now() - _code_time_9 >= 1000) {
        console.debug("[code time]demo/index.tsx:19:3: " + (Date.now() - _code_time_9) + "ms");
    }
}
var _code_memory_browser_after_2 = performance.memory.usedJSHeapSize;
var _code_memory_browser_changes_2 = _code_memory_browser_after_2 - _code_memory_browser_2;
if (_code_memory_browser_changes_2 > 10485.76) {
    console.debug("[code memory browser]demo/index.tsx:15:1: +" + Math.round(_code_memory_browser_changes_2 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_2 / 10485.76) / 100) + "MB");
}
if (Date.now() !== _code_time_2) {
    console.debug("[code time]demo/index.tsx:15:1: " + (Date.now() - _code_time_2) + "ms");
}
_lastExecutingCode = "demo/index.tsx:26:1";
_executedCodeLines.add(26);
/**
 * executed-code:disable
 * code-time:disable
 */
function foo() {
    _lastExecutingCode = "demo/index.tsx:27:3 foo";
    var _code_memory_browser_12 = performance.memory.usedJSHeapSize;
    _executedCodeLines.add(27);
    console.info(1);
    var _code_memory_browser_after_12 = performance.memory.usedJSHeapSize;
    var _code_memory_browser_changes_12 = _code_memory_browser_after_12 - _code_memory_browser_12;
    if (_code_memory_browser_changes_12 > 10485.76) {
        console.debug("[code memory browser]demo/index.tsx:27:3 foo: +" + Math.round(_code_memory_browser_changes_12 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_12 / 10485.76) / 100) + "MB");
    }
}
console.debug("[executed code]demo/index.tsx:30:1");
_lastExecutingCode = "demo/index.tsx:30:1";
_executedCodeLines.add(30);
function bar() {
    console.debug("[executed code]demo/index.tsx:31:3 bar");
    _lastExecutingCode = "demo/index.tsx:31:3 bar";
    _executedCodeLines.add(31);
    function baz() {
        console.debug("[executed code]demo/index.tsx:32:5 baz");
        var _code_time_11 = Date.now();
        _lastExecutingCode = "demo/index.tsx:32:5 baz";
        var _code_memory_browser_14 = performance.memory.usedJSHeapSize;
        _executedCodeLines.add(32);
        console.info(2);
        var _code_memory_browser_after_14 = performance.memory.usedJSHeapSize;
        var _code_memory_browser_changes_14 = _code_memory_browser_after_14 - _code_memory_browser_14;
        if (_code_memory_browser_changes_14 > 10485.76) {
            console.debug("[code memory browser]demo/index.tsx:32:5 baz: +" + Math.round(_code_memory_browser_changes_14 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_14 / 10485.76) / 100) + "MB");
        }
        if (Date.now() - _code_time_11 >= 1000) {
            console.debug("[code time]demo/index.tsx:32:5 baz: " + (Date.now() - _code_time_11) + "ms");
        }
    }
    console.debug("[executed code]demo/index.tsx:34:3 bar");
    var _code_time_10 = Date.now();
    _lastExecutingCode = "demo/index.tsx:34:3 bar";
    var _code_memory_browser_13 = performance.memory.usedJSHeapSize;
    _executedCodeLines.add(34);
    console.info(baz);
    var _code_memory_browser_after_13 = performance.memory.usedJSHeapSize;
    var _code_memory_browser_changes_13 = _code_memory_browser_after_13 - _code_memory_browser_13;
    if (_code_memory_browser_changes_13 > 10485.76) {
        console.debug("[code memory browser]demo/index.tsx:34:3 bar: +" + Math.round(_code_memory_browser_changes_13 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_13 / 10485.76) / 100) + "MB");
    }
    if (Date.now() - _code_time_10 >= 1000) {
        console.debug("[code time]demo/index.tsx:34:3 bar: " + (Date.now() - _code_time_10) + "ms");
    }
}
console.debug("[executed code]demo/index.tsx:37:1");
_lastExecutingCode = "demo/index.tsx:37:1";
_executedCodeLines.add(37);
class A {
    b() {
        console.debug("[executed code]demo/index.tsx:39:5 A.b");
        var _code_time_12 = Date.now();
        _lastExecutingCode = "demo/index.tsx:39:5 A.b";
        var _code_memory_browser_15 = performance.memory.usedJSHeapSize;
        _executedCodeLines.add(39);
        console.info(1);
        var _code_memory_browser_after_15 = performance.memory.usedJSHeapSize;
        var _code_memory_browser_changes_15 = _code_memory_browser_after_15 - _code_memory_browser_15;
        if (_code_memory_browser_changes_15 > 10485.76) {
            console.debug("[code memory browser]demo/index.tsx:39:5 A.b: +" + Math.round(_code_memory_browser_changes_15 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_15 / 10485.76) / 100) + "MB");
        }
        if (Date.now() - _code_time_12 >= 1000) {
            console.debug("[code time]demo/index.tsx:39:5 A.b: " + (Date.now() - _code_time_12) + "ms");
        }
    }
    constructor() {
        console.debug("[executed code]demo/index.tsx:42:5 A.constructor");
        var _code_time_13 = Date.now();
        _lastExecutingCode = "demo/index.tsx:42:5 A.constructor";
        var _code_memory_browser_16 = performance.memory.usedJSHeapSize;
        _executedCodeLines.add(42);
        console.info(2);
        var _code_memory_browser_after_16 = performance.memory.usedJSHeapSize;
        var _code_memory_browser_changes_16 = _code_memory_browser_after_16 - _code_memory_browser_16;
        if (_code_memory_browser_changes_16 > 10485.76) {
            console.debug("[code memory browser]demo/index.tsx:42:5 A.constructor: +" + Math.round(_code_memory_browser_changes_16 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_16 / 10485.76) / 100) + "MB");
        }
        if (Date.now() - _code_time_13 >= 1000) {
            console.debug("[code time]demo/index.tsx:42:5 A.constructor: " + (Date.now() - _code_time_13) + "ms");
        }
    }
    get c() {
        console.debug("[executed code]demo/index.tsx:45:5 A.c");
        _lastExecutingCode = "demo/index.tsx:45:5 A.c";
        _executedCodeLines.add(45);
        return 1;
    }
    set c(value) {
        console.debug("[executed code]demo/index.tsx:48:5 A.c");
        var _code_time_14 = Date.now();
        _lastExecutingCode = "demo/index.tsx:48:5 A.c";
        var _code_memory_browser_17 = performance.memory.usedJSHeapSize;
        _executedCodeLines.add(48);
        console.info(value);
        var _code_memory_browser_after_17 = performance.memory.usedJSHeapSize;
        var _code_memory_browser_changes_17 = _code_memory_browser_after_17 - _code_memory_browser_17;
        if (_code_memory_browser_changes_17 > 10485.76) {
            console.debug("[code memory browser]demo/index.tsx:48:5 A.c: +" + Math.round(_code_memory_browser_changes_17 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_17 / 10485.76) / 100) + "MB");
        }
        if (Date.now() - _code_time_14 >= 1000) {
            console.debug("[code time]demo/index.tsx:48:5 A.c: " + (Date.now() - _code_time_14) + "ms");
        }
    }
}
console.debug("[executed code]demo/index.tsx:53:1");
var _code_time_3 = Date.now();
_lastExecutingCode = "demo/index.tsx:53:1";
var _code_memory_browser_3 = performance.memory.usedJSHeapSize;
_executedCodeLines.add(53);
_showCodeCoverage();
var _code_memory_browser_after_3 = performance.memory.usedJSHeapSize;
var _code_memory_browser_changes_3 = _code_memory_browser_after_3 - _code_memory_browser_3;
if (_code_memory_browser_changes_3 > 10485.76) {
    console.debug("[code memory browser]demo/index.tsx:53:1: +" + Math.round(_code_memory_browser_changes_3 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_3 / 10485.76) / 100) + "MB");
}
if (Date.now() !== _code_time_3) {
    console.debug("[code time]demo/index.tsx:53:1: " + (Date.now() - _code_time_3) + "ms");
}
console.debug("[executed code]demo/index.tsx:55:1");
var _code_time_4 = Date.now();
_lastExecutingCode = "demo/index.tsx:55:1";
var _code_memory_browser_4 = performance.memory.usedJSHeapSize;
_executedCodeLines.add(55);
React.createElement("button", { onClick: () => {
        console.debug("[executed code]demo/index.tsx:55:24");
        return console.info(1);
    }, "data-_position": 'demo/index.tsx:55:1' });
var _code_memory_browser_after_4 = performance.memory.usedJSHeapSize;
var _code_memory_browser_changes_4 = _code_memory_browser_after_4 - _code_memory_browser_4;
if (_code_memory_browser_changes_4 > 10485.76) {
    console.debug("[code memory browser]demo/index.tsx:55:1: +" + Math.round(_code_memory_browser_changes_4 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_4 / 10485.76) / 100) + "MB");
}
if (Date.now() !== _code_time_4) {
    console.debug("[code time]demo/index.tsx:55:1: " + (Date.now() - _code_time_4) + "ms");
}
console.debug("[executed code]demo/index.tsx:57:1");
var _code_time_5 = Date.now();
_lastExecutingCode = "demo/index.tsx:57:1";
var _code_memory_browser_5 = performance.memory.usedJSHeapSize;
_executedCodeLines.add(57);
const a = 3;
var _code_memory_browser_after_5 = performance.memory.usedJSHeapSize;
var _code_memory_browser_changes_5 = _code_memory_browser_after_5 - _code_memory_browser_5;
if (_code_memory_browser_changes_5 > 10485.76) {
    console.debug("[code memory browser]demo/index.tsx:57:1: +" + Math.round(_code_memory_browser_changes_5 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_5 / 10485.76) / 100) + "MB");
}
if (Date.now() !== _code_time_5) {
    console.debug("[code time]demo/index.tsx:57:1: " + (Date.now() - _code_time_5) + "ms");
}
console.debug("[executed code]demo/index.tsx:58:1");
var _code_time_6 = Date.now();
_lastExecutingCode = "demo/index.tsx:58:1";
var _code_memory_browser_6 = performance.memory.usedJSHeapSize;
_executedCodeLines.add(58);
console.info(a);
var _code_memory_browser_after_6 = performance.memory.usedJSHeapSize;
var _code_memory_browser_changes_6 = _code_memory_browser_after_6 - _code_memory_browser_6;
if (_code_memory_browser_changes_6 > 10485.76) {
    console.debug("[code memory browser]demo/index.tsx:58:1: +" + Math.round(_code_memory_browser_changes_6 / 10485.76) / 100 + "MB -> " + (Math.round(_code_memory_browser_after_6 / 10485.76) / 100) + "MB");
}
if (Date.now() !== _code_time_6) {
    console.debug("[code time]demo/index.tsx:58:1: " + (Date.now() - _code_time_6) + "ms");
}

/* WEBPACK VAR INJECTION */}.call(this, __webpack_require__(1)))

/***/ }),
/* 1 */
/***/ (function(module, exports) {

var g;

// This works in non-strict mode
g = (function() {
	return this;
})();

try {
	// This works if eval is allowed (see CSP)
	g = g || new Function("return this")();
} catch (e) {
	// This works if the window reference is available
	if (typeof window === "object") g = window;
}

// g can still be undefined, but nothing to do about it...
// We return undefined, instead of nothing here, so it's
// easier to handle this case. if(!global) { ...}

module.exports = g;


/***/ })
/******/ ]);